draft-ietf-oauth-browser-based-apps.md

```diff
@@ -919,14 +920,22 @@ Authorization servers SHOULD link the lifetime of the refresh token to the user'
 
-For the authorization server, a proper CORS configuration is relevant for the token endpoint, where the browser-based application exchanges the authorization code for tokens. Additionally, if the authorization server provides additional endpoints to the application, such as metadata URLs, dynamic client registration, revocation, introspection, discovery or user info endpoints, these endpoints may also be accessed by the browser-based application. Consequentially, the authorization server is responsible for enforcing a proper CORS configuration on these endpoints.
+For the authorization server, a proper CORS configuration is relevant for the token endpoint, where the browser-based application exchanges the authorization code for tokens. Additionally, if the authorization server provides additional endpoints to the application, such as discovery metadata URLs, JSON Web Key Sets, dynamic client registration, revocation, introspection or user info endpoints, these endpoints may also be accessed by the browser-based application. Consequentially, the authorization server is responsible for enforcing a proper CORS configuration on these endpoints.
 
 This specification does not include guidelines for deciding whether a CORS policy
-for the token endpoint should be a wildcard origin or more restrictive. Note,
-however, that the browser will attempt to GET or POST to the API endpoint before
-knowing any CORS policy; it simply hides the succeeding or failing result from
+for these endpoints should be a wildcard origin or more restrictive. Note,
+however, that the browser will typically do a "preflight" request to checks to see
+if the CORS protocol is understood and what the policy is. In absence of support for
+preflight requests at the server the browser will attempt to access the API endpoint
+regardless and will simply hide the succeeding or failing result from
 JavaScript if the policy does not allow sharing.
 
+Note that the server CORS preflight response and the actual response CORS headers may differ,
+this allows the server to indicate wildcard support for accessing the endpoints
+during preflight when it has no access to the actual request's body or headers and then when
+the actual request is made decide to restrict the response after it was able to check the request
+was made from an expected origin based on the data in the request, e.g. based on pre-registered
+origins of the Browser-based OAuth 2.0 client.
```

Comment:
I would actually suggest to include a guideline here which is to

- have Origin allow-list pre-registerable or even required for browser based applications
- implement preflights that are wildcard and/or limited
- figure out the client_id from the actual request (client_id in body, aud/azp in tokens) and check the origin matches one that is pre-registered, otherwise error the response and/or remove the actual responses' CORS headers to have the browser hide the response properties.

Reply:
That sounds like it will require a lot of additional text to justify those recommendations.